analyzers/src/SonarAnalyzer.CSharp/Rules/CheckArgumentException.cs

```diff
@@ -100,61 +95,44 @@ private static void CheckForIssue(SyntaxNodeAnalysisContext analysisContext)
                 return;
             }
 
-            var methodArgumentNames = GetMethodArgumentNames(objectCreationSyntax);
-            if (methodArgumentNames.Contains(TakeOnlyBeforeDot(parameterNameValue)))
-            {
-                // paramName exists
-                return;
-            }
-
-            if (messageValue.HasValue &&
-                messageValue.Value != null &&
-                methodArgumentNames.Contains(TakeOnlyBeforeDot(messageValue)))
+            var methodArgumentNames = GetMethodArgumentNames(objectCreationSyntax).ToHashSet();
+            if (!methodArgumentNames.Contains(TakeOnlyBeforeDot(parameterNameValue)))
             {
-                analysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(rule,
-                    objectCreationSyntax.GetLocation(), ConstructorParametersInverted));
-                return;
+                var message = messageValue.HasValue && messageValue.Value != null && methodArgumentNames.Contains(TakeOnlyBeforeDot(messageValue))
+                    ? ConstructorParametersInverted
+                    : string.Format(InvalidParameterName, parameterNameValue.Value);
+                analysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, objectCreationSyntax.GetLocation(), message));
             }
-
-            analysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(rule, objectCreationSyntax.GetLocation(),
-                    string.Format(InvalidParameterName, parameterNameValue.Value)));
         }
 
-        private static ISet<string> GetMethodArgumentNames(ObjectCreationExpressionSyntax creationSyntax)
+        private static IEnumerable<string> GetMethodArgumentNames(ObjectCreationExpressionSyntax creationSyntax)
```

Comment:
That's one very long method. I think it would be better to split it.

Reply:
I intentionally joined them, because there was hidden dependency between them. The first part lists some types and the second one processes that exact list of types.

Reply:
And instead of having one very long method we could extract one method `GetMethodDeclaration` (responsibility will be clear, there will be one single reason for change in this case) and `GetIdentifierNames` responsible for providing the names. The second can call directly the first or receive the SyntaxNode as parameter. One other thing we can improve is the second part which seems to be replaceable with a switch expression or statement.

Reply:
Two methods like this was the original state. While they were easy to read individually, the hidden dependency made them hard to work with.

The switch expression removed quite a lot of noise. Please take a look if if fixes the readability issue.

Reply:
I think it's certainly better since it reduced a lot the size of the method 😄